### PR TITLE
Plans 2023: Typed Async props and other prop cleanup

### LINKED ISSUE #78572: Plans 2023: Properly type the async-props passed from PlansFeaturesMain to the plans grid components
As of https://github.com/Automattic/wp-calypso/pull/78428 the `PlansFeaturesMain` wrapper is properly typed. However, the grid components from `plan-features-2023-grid` are loaded via Async, and the props passed through are not typed/matched to `plan-features-2023-grid` props. This will likely lead to deprecations (things passed through but not used) as the two are aligned e.g. `popularPlanSpec` (and all the computations around it) seem subject to deletion.

- Export the `PlanFeatures2023GridProps` from `plan-features-2023-grid`
- Use it for typing the props passed to the Async component
- Fix the mismatches. IMPORTANT: use `PlansFeaturesMain` as the source of truth for what is optional/required, where relevant

### PR DESCRIPTION
**Pull request by chriskmnds**
Fixes https://github.com/Automattic/wp-calypso/issues/78572

## Proposed Changes

- Types the async props passed to `plan-features-2023-grid` form `plans-features-main` (gotta love these naming conventions)
- Removes several redundant/unused properties
- Several TODOs added to be picked up (by author/me) in other work

## Testing Instructions

* Confirm `/start/plans` and `/plans/[paid and free]` work as before. 
* Try a tailored flow e.g. `/setup/link-in-bio`
* Try `plans/[woo site]` as that imports `plan-features-2023-grid` directly
* Notices, plan prices, features, etc. show as normal

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts]: note to self: probably turn to ternary, ensure `undefined` return

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/index.tsx]: note to self: probably address this here

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/index.tsx:241]: there will be some overhaul in this file next

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/index.tsx:909]: note to self: convert to "plans-package" issue

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/index.tsx:914]: note to self: convert to a separate issue. there's a bit of drilling involved

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/index.tsx:917]: note to self: convert to issue if not part of one already

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: separate issue coming. this should probably be refactored out instead

**Comment by chriskmnds** [comment]
p.s. all my comments/TODOs concerning issues to create, etc, are resolved. a lot of issues will be put on the board soon

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/index.tsx]: Nit. 
```suggestion
	hideUnavailableFeatures?: boolean; // used to hide features that are not available, instead of strike-through as explained in #76206
```

File: client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts:24]: Do we need to handle the siteId empty cases here? That is handled already down the tree inside the selector.

i.e. 
https://github.com/Automattic/wp-calypso/blob/79b9877d76cbd8b424d0b7940c6470cb22841b67/client/state/sites/selectors/get-site-plan.ts#L30-L32

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/index.tsx:634]: This comment is a bit unclear. Because `ownPropsOnUpgradeClick?.( cartItemForPlan )` is already used 🤔

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/index.tsx]: This is not related to this PR. But the actual fix should happen here.

https://github.com/Automattic/wp-calypso/blob/79b9877d76cbd8b424d0b7940c6470cb22841b67/packages/onboarding/src/utils/flows.ts#L88

----

To Elaborate ...

The `string | null` type convention for the flowName that can be seen  in the file [flows.ts](https://github.com/Automattic/wp-calypso/blob/trunk/packages/onboarding/src/utils/flows.ts) was introduced first in the following commit.
https://github.com/Automattic/wp-calypso/blob/ca4fce1d425cdfd11f4d41cc93b30f376322baf7/packages/onboarding/src/utils.ts#L1-L6 

Since then the convention has been growing! However, within the implementation, none of those functions are actually checking for the null value. It is just checking for coerced truthy value.

https://github.com/Automattic/wp-calypso/blob/79b9877d76cbd8b424d0b7940c6470cb22841b67/packages/onboarding/src/utils/flows.ts#L84-L97

We should ideally fix that to communicate reality but I do understand that it is not the scope of this PR. Just noticing how small things can grow in unanticipated ways.

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/index.tsx:1045]: Nice 👌

File: client/my-sites/plans-features-main/index.tsx
Comment (ddc22) [inline client/my-sites/plans-features-main/index.tsx:140]: Unrelated to PR and Maybe you have already thought of this but. Could we possibly pass `planActionOverrides` via the `PlansGridContext`

File: client/my-sites/plans-features-main/index.tsx
Comment (ddc22) [inline client/my-sites/plans-features-main/index.tsx]: To avoid this cast we should probably use `import { TranslateResult } from 'i18n-calypso';` below.

https://github.com/Automattic/wp-calypso/blob/72b5cc24e46f5205c234168ca8c85485d2b2195b/client/my-sites/plan-features-2023-grid/types.ts#L41

File: client/my-sites/plan-features-2023-grid/hooks/use-calculate-max-plan-upgrade-credit.ts
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/hooks/use-calculate-max-plan-upgrade-credit.ts:11]: Nit. We could probably include this inline.

File: client/my-sites/plan-features-2023-grid/components/features.tsx
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/components/features.tsx]: Should `domainName` empty check also include the else part of this condition?

https://github.com/Automattic/wp-calypso/blob/907954e4dc25dd51b4e26a95bc07f7c7ed8c3b4a/client/my-sites/plan-features-2023-grid/components/features.tsx#L94-L98

**Review by ddc22 — COMMENTED** [review]
I have left some code comments below. Will do some testing next.

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: 💯 Thank you (also saved me some trouble with other PR)

I wonder if a better way/name to make this a little more direct. Somehow the opposite strikes me as more visible `showUnavalableFeatures`

File: client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/hooks/use-is-plan-upgrade-credit-visible.ts:24]: Not sure I follow. The change is addressing a type mismatch with `getSitePlanSlug`

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx:634]: Hmm old comment. Probably on the wrong premise, but I left it here to catch on another refactor. I think it refers to `if ( cartItemForPlan ) { ...`

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: I'm not sure I see the problem fully as you do here. The issue is with `isAnyHostingFlow` not accepting undefined for flowName (almost like treating `null` as the "optional" part of a prop definition). I'd expect the type to be `flowName?: string | null` if we want to allow treatment of empty values.

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx:140]: Yeah, we'll come to the grid context at a later point and decide what to put in there. It is a somewhat suboptimal way of passing props around e.g. a change to a context prop will force rerender anything consuming the entire context. 

We can explore a data store instead to hold the grid props e.g. some onLoad hook to populate a store the props passed. that should save us with all the prop drilling like a context would. Offhand, haven't really thought about it. context might just do.

File: client/my-sites/plan-features-2023-grid/hooks/use-calculate-max-plan-upgrade-credit.ts
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/hooks/use-calculate-max-plan-upgrade-credit.ts:11]: I prefer to have `interface Props { ...` for anything that's in some standalone and exported form.

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: we most certainly should. thank you 💯

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (ddc22) [reply client/my-sites/plan-features-2023-grid/index.tsx]: Sorry that is exactly what I meant 😄 ... That ( `flowName?: string | null` ) should have been the prop from the get go for every single `is...` call in that file!

File: client/my-sites/plan-features-2023-grid/components/features.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/components/features.tsx]: I don't think it should. It will never reach that point. This is largely the result of returning early on line 59 above:

```
if ( isFreePlanAndCustomDomainFeature && ! domainName ) {
	return null;
}
```

and following with `isFreePlanAndCustomDomainFeature` and using `domainName` within - it creates a type mismatch. One of the reasons I do not like early returns on complex conditions like this. It confuses the rest of the code.

Would you prefer casting `domainName` to string at point of use? (not sure if I do personally)

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: sound. I'll make the change and let the rest follow

File: client/my-sites/plan-features-2023-grid/components/features.tsx
Comment (ddc22) [reply client/my-sites/plan-features-2023-grid/components/features.tsx]: Oh right I have seen this annoyance as well 😬.

> This is largely the result of returning early on line 59 above:

I did not notice this ☝️ .

> Would you prefer casting domainName to string at point of use? (not sure if I do personally)

This would be less confusing for someone looking at the code later, and a legitimate use of `as` since the typescript compiler doesn't seem to be picking up on the correct type here.

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: should be done

**Comment by chriskmnds** [comment]
Thanks for the review @jdc91 . Pretty valuable comments. Love how you pinpoint things sometimes. I've addressed and resolved what I feel should be addressed right now and left some follow-up comments to the rest/unresolved.

File: client/my-sites/plan-features-2023-grid/components/features.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/components/features.tsx]: yeah, that sounds right. updated

**Comment by ddc22** [comment]
Sadly I could not follow through with testing. But generally the code LGTM. @aneeshd16 or @jeyip  will you have some time to go over the testing for this (and even more code review if required 😁). Don't want to leave @chriskmnds waiting for an approval 😶‍🌫️ .

**Comment by jeyip** [comment]
> @aneeshd16 or @jeyip will you have some time to go over the testing for this (and even more code review if required 😁). Don't want to leave @chriskmnds waiting for an approval 😶‍🌫️ .

I was [pulled into work with the Bumblebees](https://github.com/Automattic/martech/issues/1907) and just saw this comment at the end of my day -- I can take a look first thing tomorrow morning if Aneesh hasn't gotten to it 🙂

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (aneeshd16) [inline client/my-sites/plan-features-2023-grid/index.tsx]: We can probably remove this redundant comment.

**Review by aneeshd16 — APPROVED** [review]
LGTM! Tested these flows:

* `/start`
* `/start/launch-site`
* `/plans/`
* `/setup/newsletter/plans`
* `domainAndPlanPackage` flow

**Comment by jeyip** [comment]
## Testing
### Requirements

Did a visual smoke test of the Pricing Feature Grid ( and Plans Comparison Grid where possible ) for the following flows:

- [x] /start/plans with free plan
- [x] /plans with free plan and paid plan
- [x] /setup/link-in-bio
- [x] /setup/wooexpress ( looked at the plans page )

Misc:

- [x] Can upgrade plans as expected
- [x] Can purchase plans as expected 

### Browsers
- [x] Chrome

**Review by jeyip — APPROVED** [review]

File: client/my-sites/plan-features-2023-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/index.tsx]: we most certainly should!

**Pull request merged**
